Add Date aggregation {wip} (#1585)
enkidevs/curriculum: mongodb/intro/date-aggregation/aggregation-with-date-expression-operators.md

---
author: Stefan-Stojanovic

levels:
  - beginner
  - basic
  
aspects:
  - introduction
  - new

type: normal

category: how to

---

# Intro to Aggregation With Date Expression Operators

---
## Content

In MongoDB, it is possible to aggregate a collection using date expression operators. These operators can be used to extract the specific time at which a document was made, they can be used to find a group of documents created on a specific date or to find all documents between 2 dates.

There are a total of 10 operators that can be used to aggregate a collection by date.

These are:
- `$year`
- `$month`
- `$dayOfMonth`
- `$week`
- `$hour`
- `$minute`
- `$second`
- `$milisecond`
- `$dayOfYear`
- `$dayOfWeek`

In a previous workout, we learned what `ObjectId`s are and how to extract the date in the ISODate format. 

Short reminder:

> Let's say we have this pokémon(document).

Reviewer: ```suggestion
Let's say we have this pokémon document.
```

```javascript
{ 
  "_id": ObjectId("5d9d8a6a0b24990f19398209"),
  "name": "Bulbasaur",
  "type": "Grass"
}
```

We can copy the ObjectId: `ObjectId("5d9d8a6a0b24990f19398209")` and add the `.getTimestamp()` method after it in our `mongodb` shell and hit enter.

This will output the date part from the `ObjectId` in an ISODate format.

Output:
```javascript
ISODate("2019-10-09T07:21:14Z")

> ```

Reviewer: Do you think it would be clearer to 1) remove the sample document that doesn't have the date field from all footnotes 2) add the sample document that has the date field here and say it will be used in the upcoming insights

```md
To best demonstrate the usage of various date expression operators, we will add a creation date to our `"Bulbasaur"` pokémon and use it as the example in the upcoming insights:

```json
{ 
  "_id": ObjectId("5d9d8a6a0b24990f19398209"),
  "name": "Bulbasaur",
  "type": "Grass",
  "Date": ISODate("2019-10-09T07:21:14Z"),
  // "createdAt": ISODate("2019-10-09T07:21:14Z") <-- should use this one due to inconsistent capitalization
}
```

Author: The "problem" with changing it to `createdAt` I need to go change every example, + check the database I used and change it there as well In case I need to use it again. Should I change it everywhere?

Reviewer 2: This shouldn't take too long to do.

On the content side, you can use the find & replace feature of vs code.

On the db side, you can use the [`$rename` operator](https://docs.mongodb.com/manual/reference/operator/update/rename/#rename-a-field) to change the name of the field.

---
## Practice

Which method is used to extract the date from an `ObjectId` value in an ISODate format?

???

* `getTimestamp()`
* `getDate()`
* `getTime()`

> * `timeStamp()`

Reviewer 2: ```suggestion
* `timeStamp()`

```

---
author: Stefan-Stojanovic

aspects:
  - introduction
  - new

type: normal

category: how to

---

# Intro to Aggregation With Date Expression Operators

---
## Content

> In MongoDB, it is possible to aggregate a collection using date expression operators. These operators can be used to extract the specific time at which a document was made, they can be used to find a group of documents created on a specific date or to find all documents between 2 dates.

Reviewer 2: ```suggestion
In MongoDB, it is possible to aggregate a collection using date expression operators. 

These operators can be used to extract the specific time at which a document was made, find a group of documents created on a specific date, or to find all documents between two dates.
```

> There are a total of 10 operators that can be used to aggregate a collection by date.
> 
> These are:
> - `$year`
> - `$month`
> - `$dayOfMonth`
> - `$week`
> - `$hour`
> - `$minute`
> - `$second`
> - `$milisecond`
> - `$dayOfYear`
> - `$dayOfWeek`

Reviewer 2: ```suggestion
There are a total of 10 operators that can be used to aggregate a collection by date:
- `$year`
- `$month`
- `$dayOfMonth`
- `$week`
- `$hour`
- `$minute`
- `$second`
- `$milisecond`
- `$dayOfYear`
- `$dayOfWeek`
```
